pep-0573.rst

```diff
@@ -385,30 +394,51 @@ and return -1.
 If called with an initialized exception type (``*exc``
 is non-NULL), the function will do nothing but incref ``*exc``.
 
-(XXX: Support and test multiple inheritance)
-
 A new flag, ``Py_TPFLAGS_HEAP_IMMUTABLE``, will be added to prevent
 mutation of the type object. This makes it possible to
 share the object safely between multiple interpreters.
-This flag is checked in setattr (XXX: name of C function?) and blocks
+This flag is checked in ``type_setattro`` and blocks
 setting of attributes when set, similar to built-in types.
 
 A new pointer, ``ht_moduleptr``, will be added to heap types to store ``exc``.
 
-(XXX: better name? Update implementation)
-
 On deinitialization of the exception type, ``*exc`` will be set to ``NULL``.
-This makes it safe for
-``PyErr_PrepareImmutableException`` to check if the exception
-was already initialized.
+This makes it safe for ``PyErr_PrepareImmutableException`` to check if
+the exception was already initialized.
+
+PyType_offsets
+--------------
+
+Some extension types are using instances with ``__dict__`` or ``__weakref__``
+allocated. Currently, there is no way of passing offsets of these through
+``PyType_Spec``. To allow this, a new structure and a spec slot are proposed.
+
+A new structure, ``PyType_offsets``, will have two members containing the
+offsets of ``__dict__`` and ``__weakref__``::
+
+    typedef struct {
+        Py_ssize_t dict;
+        Py_ssize_t weaklist;
+    } PyType_offsets;
+
+The new slot, ``Py_offsets``, will be used to pass a ``PyType_offsets *``
+structure containing the mentioned data.
```

Comment:
Due to the stable ABI restrictions, it would be better if these were two separate slots - that way, if we need a new offset in the future, it would just be another slot definition.

If we use a struct, we can never make it bigger, as that could lead to a newer Python reading past the end of the memory allocated by an older extension module.